drivers/entropy/entropy_mspm0_trng.c

```diff
@@ -0,0 +1,300 @@
+	DL_TRNG_setDecimationRate(config->base, (DL_TRNG_DECIMATION_RATE)TRNG_DECIMATION_RATE);
+
+	DL_TRNG_sendCommand(config->base, DL_TRNG_CMD_NORM_FUNC);
+	wait_for_command_done(config);
+
+	DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CMD_DONE_EVENT);
+
+	/* Enable health fail interrupt */
+	DL_TRNG_enableInterrupt(config->base, DL_TRNG_INTERRUPT_HEALTH_FAIL_EVENT);
+
+	/* Enable data captured interrupt */
+	DL_TRNG_enableInterrupt(config->base, DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT);
+
+	dummy_data = DL_TRNG_getCapture(config->base);
+	DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT);
+
+	return 0;
+}
+
+static void entropy_mspm0_trng_isr(const struct device *dev)
+{
+	const struct entropy_mspm0_trng_config *config = dev->config;
+	struct entropy_mspm0_trng_data *data = dev->data;
+	uint32_t status;
+	uint32_t entropy_data;
+
+	status = DL_TRNG_getRawInterruptStatus(config->base,
+		 DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT | DL_TRNG_INTERRUPT_HEALTH_FAIL_EVENT);
+
+	if (status & DL_TRNG_INTERRUPT_HEALTH_FAIL_EVENT) {
+		DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_HEALTH_FAIL_EVENT);
+		return;
+	}
+
+	if ((status & DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT) &&
+	     DL_TRNG_isCaptureReady(config->base)) {
+		entropy_data = DL_TRNG_getCapture(config->base);
+		ring_buf_put(&data->entropy_pool, (uint8_t *)&entropy_data, TRNG_SAMPLE_SIZE);
+		DL_TRNG_clearInterruptStatus(config->base, DL_TRNG_INTERRUPT_CAPTURE_RDY_EVENT);
+		k_sem_give(&data->sem_sync);
+	}
+}
+
+static int entropy_mspm0_trng_get_entropy(const struct device *dev,
+					  uint8_t *buffer, uint16_t length)
+{
+	struct entropy_mspm0_trng_data *data = dev->data;
+	uint16_t bytes_read;
+
+	if (!data->initialized) {
+		return -ENODEV;
+	}
+
+	while (length) {
+		k_sem_take(&data->sem_lock, K_FOREVER);
+		bytes_read = ring_buf_get(&data->entropy_pool, buffer, length);
+		k_sem_give(&data->sem_lock);
+
+		if (bytes_read) {
+			buffer += bytes_read;
+			length -= bytes_read;
+		} else {
+			k_sem_take(&data->sem_sync, K_FOREVER);
+		}
+	}
+
+	return 0;
+}
+
+static int entropy_mspm0_trng_get_entropy_isr(const struct device *dev, uint8_t *buffer,
+					      uint16_t length, uint32_t flags)
+{
+	const struct entropy_mspm0_trng_config *config = dev->config;
+	struct entropy_mspm0_trng_data *data = dev->data;
+	uint16_t bytes_read = 0;
+	uint16_t total_read = 0;
+	uint32_t entropy_data;
+	unsigned int key;
+
+	if (!data->initialized) {
+		return -ENODEV;
+	}
+
+	/* Try to get entropy from existing ring buffer */
+	bytes_read = ring_buf_get(&data->entropy_pool, buffer, length);
+
+	total_read = bytes_read;
+
+	if ((bytes_read == length) || ((flags & ENTROPY_BUSYWAIT) == 0U)) {
+		/* Either we got all requested data, or busy-waiting is not allowed */
+		return total_read;
+	}
+
+	/* Busy-wait for additional data (only if ENTROPY_BUSYWAIT is set) */
+	buffer += bytes_read;
+	length -= bytes_read;
+
+	while (length) {
+		key = irq_lock();
```

Comment:
this fp is meant to be called from ISR already, which means CPU is already in ISR context. Do we to stall the remaining interrupts of the CPU?